// @massalabs/massa-web3/dist/cmd/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.utils = exports.MassaUnits = exports.toMAS = exports.fromMAS = exports.PROTO_FILE_SEPARATOR = exports.MASSA_PROTOFILE_KEY = exports.SmartContractsClient = exports.ON_MASSA_EVENT_ERROR = exports.ON_MASSA_EVENT_DATA = exports.EventPoller = exports.WalletClient = exports.PrivateApiClient = exports.PublicApiClient = exports.Client = exports.DefaultProviderUrls = exports.ClientFactory = exports.OperationTypeId = exports.EOperationStatus = exports.ProviderType = void 0;
const tslib_1 = require("tslib");
/** Polyfills */
const buffer_1 = require("buffer");
const events_1 = tslib_1.__importDefault(require("events"));
// Check if we are on browser
if (typeof window !== 'undefined') {
    window.Buffer = buffer_1.Buffer;
    window.EventEmitter = events_1.default;
}
var IProvider_1 = require("./interfaces/IProvider");
Object.defineProperty(exports, "ProviderType", { enumerable: true, get: function () { return IProvider_1.ProviderType; } });
var EOperationStatus_1 = require("./interfaces/EOperationStatus");
Object.defineProperty(exports, "EOperationStatus", { enumerable: true, get: function () { return EOperationStatus_1.EOperationStatus; } });
var OperationTypes_1 = require("./interfaces/OperationTypes");
Object.defineProperty(exports, "OperationTypeId", { enumerable: true, get: function () { return OperationTypes_1.OperationTypeId; } });
/** Exposed clients and factories */
var ClientFactory_1 = require("./web3/ClientFactory");
Object.defineProperty(exports, "ClientFactory", { enumerable: true, get: function () { return ClientFactory_1.ClientFactory; } });
Object.defineProperty(exports, "DefaultProviderUrls", { enumerable: true, get: function () { return ClientFactory_1.DefaultProviderUrls; } });
var Client_1 = require("./web3/Client");
Object.defineProperty(exports, "Client", { enumerable: true, get: function () { return Client_1.Client; } });
var PublicApiClient_1 = require("./web3/PublicApiClient");
Object.defineProperty(exports, "PublicApiClient", { enumerable: true, get: function () { return PublicApiClient_1.PublicApiClient; } });
var PrivateApiClient_1 = require("./web3/PrivateApiClient");
Object.defineProperty(exports, "PrivateApiClient", { enumerable: true, get: function () { return PrivateApiClient_1.PrivateApiClient; } });
var WalletClient_1 = require("./web3/WalletClient");
Object.defineProperty(exports, "WalletClient", { enumerable: true, get: function () { return WalletClient_1.WalletClient; } });
var EventPoller_1 = require("./web3/EventPoller");
Object.defineProperty(exports, "EventPoller", { enumerable: true, get: function () { return EventPoller_1.EventPoller; } });
Object.defineProperty(exports, "ON_MASSA_EVENT_DATA", { enumerable: true, get: function () { return EventPoller_1.ON_MASSA_EVENT_DATA; } });
Object.defineProperty(exports, "ON_MASSA_EVENT_ERROR", { enumerable: true, get: function () { return EventPoller_1.ON_MASSA_EVENT_ERROR; } });
var SmartContractsClient_1 = require("./web3/SmartContractsClient");
Object.defineProperty(exports, "SmartContractsClient", { enumerable: true, get: function () { return SmartContractsClient_1.SmartContractsClient; } });
Object.defineProperty(exports, "MASSA_PROTOFILE_KEY", { enumerable: true, get: function () { return SmartContractsClient_1.MASSA_PROTOFILE_KEY; } });
Object.defineProperty(exports, "PROTO_FILE_SEPARATOR", { enumerable: true, get: function () { return SmartContractsClient_1.PROTO_FILE_SEPARATOR; } });
tslib_1.__exportStar(require("@massalabs/web3-utils"), exports);
var converters_1 = require("./utils/converters");
Object.defineProperty(exports, "fromMAS", { enumerable: true, get: function () { return converters_1.fromMAS; } });
Object.defineProperty(exports, "toMAS", { enumerable: true, get: function () { return converters_1.toMAS; } });
Object.defineProperty(exports, "MassaUnits", { enumerable: true, get: function () { return converters_1.MassaUnits; } });
/**
 * This namespace provides utility functions, such as unit conversion, serialization, encoding, etc.
 */
exports.utils = tslib_1.__importStar(require("./utils"));
