// @massalabs/massa-web3/dist/cmd/web3/ClientFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientFactory = exports.DefaultProviderUrls = void 0;
const IProvider_1 = require("../interfaces/IProvider");
const Client_1 = require("./Client");
const Web3Account_1 = require("./accounts/Web3Account");
const PublicApiClient_1 = require("./PublicApiClient");
const WalletProviderAccount_1 = require("./accounts/WalletProviderAccount");
/** Global connection urls, for Massa's MAINNET, TESTNET, LABNET, LOCALNET and BUILDNET */
var DefaultProviderUrls;
(function (DefaultProviderUrls) {
    DefaultProviderUrls["MAINNET"] = "https://massa.net/api/v2";
    DefaultProviderUrls["TESTNET"] = "https://test.massa.net/api/v2";
    DefaultProviderUrls["BUILDNET"] = "https://buildnet.massa.net/api/v2";
    DefaultProviderUrls["LABNET"] = "https://labnet.massa.net/api/v2";
    DefaultProviderUrls["LOCALNET"] = "http://127.0.0.1";
})(DefaultProviderUrls = exports.DefaultProviderUrls || (exports.DefaultProviderUrls = {}));
/**
 * Massa Web3 ClientFactory class allows you to easily initialize a client to
 * connect to the massa blockchain.
 *
 * @remarks
 * The client can be initialized using a default provider (MAINNET, TESTNET, LABNET, LOCALNET and BUILDNET)
 * or a custom set of providers.
 */
class ClientFactory {
    /**
     * Creates a default client using a default provider (MAINNET, TESTNET, LABNET, LOCALNET and BUILDNET).
     *
     * @param provider - Default provider to be used by the client.
     * @param retryStrategyOn - Whether to retry failed requests.
     * @param baseAccount - Base account to use with the client (optional).
     *
     * @returns A promise that resolves to a Client object.
     */
    static async createDefaultClient(provider, retryStrategyOn = true, baseAccount) {
        let publicProviderUrl = provider.toString();
        let privateProviderUrl = provider.toString();
        switch (provider) {
            // in the case of LocalNet append specific default ports to url.
            case DefaultProviderUrls.LOCALNET: {
                privateProviderUrl = `${privateProviderUrl}:33034`;
                publicProviderUrl = `${publicProviderUrl}:33035`;
                break;
            }
            // all other networks should have public access only.
            default: {
                break;
            }
        }
        const providers = [
            {
                url: publicProviderUrl,
                type: IProvider_1.ProviderType.PUBLIC,
            },
            {
                url: privateProviderUrl,
                type: IProvider_1.ProviderType.PRIVATE,
            },
        ];
        let clientConfig = {
            retryStrategyOn,
            providers,
        };
        let publicApi = new PublicApiClient_1.PublicApiClient(clientConfig);
        let account = null;
        if (baseAccount) {
            account = new Web3Account_1.Web3Account(baseAccount, publicApi);
        }
        const client = new Client_1.Client({
            retryStrategyOn,
            providers,
        }, account, publicApi);
        return client;
    }
    /**
     * Initializes a new client using a custom set of private and public providers.
     *
     * @remarks
     * Suitable for local node interactions.
     *
     * @param providers - Array of providers to be used by the client.
     * @param retryStrategyOn - Whether to retry failed requests.
     * @param baseAccount - Base account to be used by the client (optional).
     *
     * @returns A promise that resolves to a Client object.
     */
    static async createCustomClient(providers, retryStrategyOn = true, baseAccount) {
        let clientConfig = {
            retryStrategyOn,
            providers,
        };
        let publicApi = new PublicApiClient_1.PublicApiClient(clientConfig);
        let account = null;
        if (baseAccount) {
            account = new Web3Account_1.Web3Account(baseAccount, publicApi);
        }
        const client = new Client_1.Client(clientConfig, account, publicApi);
        return client;
    }
    /**
     * Initializes a new client using the wallet provider.
     *
     * @remarks
     * Suitable for local node interactions.
     *
     * @param provider - Provider from wallet provider to be used by the client.
     * @param baseAccount - Base account from the wallet provider to be used by the client.
     * @param retryStrategyOn - Whether to retry failed requests.
     *
     * @returns A promise that resolves to a Client object.
     */
    static async fromWalletProvider(provider, baseAccount, retryStrategyOn = true) {
        const providers = (await provider.getNodesUrls()).map((url) => {
            return {
                url,
                type: IProvider_1.ProviderType.PUBLIC,
            };
        });
        const client = new Client_1.Client({
            retryStrategyOn,
            providers,
        }, new WalletProviderAccount_1.WalletProviderAccount(baseAccount));
        return client;
    }
}
exports.ClientFactory = ClientFactory;
